pkg/distributor/distributor.go

```diff
@@ -1132,6 +1133,63 @@ func (d *Distributor) wrapPushWithMiddlewares(next PushFunc) PushFunc {
 
 }
 
+type replicaState int
+
+const (
+	replicaRejectedUnknown replicaState = 0
+	replicaIsPrimary       replicaState = 1 << iota
+	replicaNotHA
+	replicaDeduped
+	replicaRejectedTooManyClusters
+
+	replicaAccepted = replicaIsPrimary | replicaNotHA
+)
+
+type haReplica struct {
+	cluster, replica string
+}
+
+type replicaInfo struct {
+	state       replicaState
+	sampleCount int
+}
+
+// replicaObserved checks if a sample from a given replica should be accepted for ingestion based on HA deduplication rules.
+//
+// Returns a replicaState indicating the acceptance status and classification of the replica:
+//   - replicaIsPrimary: sample is from the elected primary replica and should be accepted
+//   - replicaNotHA: sample doesn't have both HA labels and should be accepted
+//   - replicaDeduped: sample is from a non-primary replica and should be deduplicated
+//   - replicaRejectedTooManyClusters: sample is rejected because the tenant has too many HA clusters
+//   - replicaRejectedUnknown: sample is rejected due to an unknown error
+func (d *Distributor) replicaObserved(ctx context.Context, userID string, replica haReplica, ts int64) (replicaState, error) {
+	span := trace.SpanFromContext(ctx)
+	span.SetAttributes(
+		// Make a copy of these, since they may be retained as tags
+		attribute.String("cluster", strings.Clone(replica.cluster)),
+		attribute.String("replica", strings.Clone(replica.replica)),
+	)
+
+	isAccepted, err := d.checkSample(ctx, userID, replica.cluster, replica.replica, ts)
+	if err != nil {
+		switch {
+		case errors.As(err, &replicasDidNotMatchError{}):
+			// These samples have been deduped.
+			return replicaDeduped, err
+		case errors.As(err, &tooManyClustersError{}):
+			return replicaRejectedTooManyClusters, err
+		default:
+			return replicaRejectedUnknown, err
+		}
+	}
+
+	if isAccepted {
+		return replicaIsPrimary, nil
+	}
+	// If there wasn't an error but isAccepted is false that means we didn't find both HA labels.
+	return replicaNotHA, nil
+}
+
 func (d *Distributor) prePushHaDedupeMiddleware(next PushFunc) PushFunc {
```

Comment:
there's way too many things happening in this function already let's break it up. i'll try to leave some suggestions